### PR TITLE
bugfix(planner): Fix wrong result of hash join when join keys have different types

### PR DESCRIPTION
**Pull request by leiysky**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

The root cause is comparison between expressions with different types and nullability.

## Changelog

- Bug Fix

### CODE REVIEW

File: query/src/sql/planner/binder/join.rs [hunk 1]
```diff
@@ -229,7 +231,13 @@ impl<'a> JoinConditionResolver<'a> {
                 },
             );
 
-            // TODO(leiysky): bump types of left conditions and right conditions
+            // Bump types of left conditions and right conditions
+            let left_type = left.data_type();
+            let right_type = right.data_type();
+            let least_super_type = merge_types(&left_type, &right_type)?;
```
Comment (leiysky) [inline query/src/sql/planner/binder/join.rs:237]: @sundy-li Seems `compare_coercion` doesn't work with `Nullable` types, is that by design?

By the way, it looks have the same function with `merge_types` except the nullability check.

Comment (sundy-li) [reply query/src/sql/planner/binder/join.rs:237]: Yes， it's by design.  I'll document it that we should use `merge_types` if we should care about nullable.

File: query/src/sql/planner/binder/scalar_common.rs [hunk 2]
```diff
@@ -115,3 +118,12 @@ pub fn split_equivalent_predicate(scalar: &Scalar) -> Option<(Scalar, Scalar)> {
         _ => None,
     }
 }
+
+pub fn wrap_cast(scalar: Scalar, target_type: &DataTypeImpl) -> Scalar {
+    let cast = CastExpr {
```
Comment (sundy-li) [inline query/src/sql/planner/binder/scalar_common.rs:123]: Add if `scalar.data_type() != target_type` special case

Comment (leiysky) [reply query/src/sql/planner/binder/scalar_common.rs:123]: Is `DataType` comparable? 

At the beginning, I was trying to use `DataTypeID` to compare the two types, but found it doesn't work with nested data types.

Maybe I should compare them with type names?

Comment (sundy-li) [reply query/src/sql/planner/binder/scalar_common.rs:123]: Yes, it's comparable.

```
impl PartialEq for DataTypeImpl {
    fn eq(&self, that: &DataTypeImpl) -> bool {
        equal(self, that)
    }
}
```

File: query/src/sql/planner/binder/join.rs (hunk 1 above)
Comment (leiysky) [reply query/src/sql/planner/binder/join.rs:237]: > Yes， it's by design.  I'll document it that we should use `merge_types` if we should care about nullable.

IMHO, `get_least_super_type` is a reasonable name.

File: query/src/sql/planner/binder/scalar_common.rs (hunk 2 above)
Comment (leiysky) [reply query/src/sql/planner/binder/scalar_common.rs]: @sundy-li Fixed.

**Review by sundy-li — APPROVED** [review]

**Review by BohuTANG — APPROVED** [review]

**Comment by leiysky** [comment]
@mergify requeue

**Pull request merged**